File: kbs_bbs/bbs2www/src/bbsmailcon.c
#include "bbslib.h"

int main() {
	FILE *fp;
	char buf[512], dir[80], file[80], path[80], *ptr, *id;
	struct fileheader x;
	int num, tmp, total;
	init_all();
	strsncpy(file, getparm("file"), 32);
	num=atoi(getparm("num"));
	printf("<center>\n");
	id=currentuser->userid;
	printf("%s -- 阅读信件 [使用者: %s]<hr>\n", BBSNAME, id);
	if(strncmp(file, "M.", 2)) http_fatal("错误的参数1");
	if(strstr(file, "..") || strstr(file, "/")) http_fatal("错误的参数2");
	sprintf(dir, "mail/%c/%s/.DIR", toupper(id[0]), id);
	total=file_size(dir)/sizeof(x);
	if(total<=0) http_fatal("错误的参数3");
	printf("<table width=610 border=1>\n");
	printf("<tr><td>\n<pre>");
	sprintf(path, "mail/%c/%s/%s", toupper(id[0]), id, file);
	fp=fopen(path, "r");
	if(fp==0) http_fatal("本文不存在或者已被删除");
	while(1) {
		if(fgets(buf, 512, fp)==0) break;
		hhprintf("%s", void1(buf));
	}
	fclose(fp);
	printf("</pre>\n</table><hr>\n");
	printf("[<a onclick='return confirm(\"你真的要删除这封信吗?\")' href=bbsdelmail?file=%s> 删除</a>]", file);
	fp=fopen(dir, "r+");
	if(fp==0) http_fatal("dir error2");
	if(num>0) {
		fseek(fp, sizeof(x)*(num-1), SEEK_SET);
		fread(&x, sizeof(x), 1, fp);
		printf("[<a href=bbsmailcon?file=%s&num=%d>上一篇</a>]", x.filename, num-1);
	}
	printf("[<a href=bbsmail>返回信件列表</a>]");
	if(num<total-1) {
		fseek(fp, sizeof(x)*(num+1), SEEK_SET);
      		fread(&x, sizeof(x), 1, fp);
      		printf("[<a href=bbsmailcon?file=%s&num=%d>下一篇</a>]", x.filename, num+1);
	}
	if(num>=0 && num<total) {
		char title2[80];
		fseek(fp, sizeof(x)*num, SEEK_SET);
		if(fread(&x, sizeof(x), 1, fp)>0) {
			x.accessed[0]|=FILE_READ;
			fseek(fp, sizeof(x)*num, SEEK_SET);
			fwrite(&x, sizeof(x), 1, fp);
			ptr=strtok(x.owner, " (");
			if(ptr==0) ptr=" ";
			strcpy(title2, x.title);
			if(strncmp(x.title, "Re:",3)) sprintf(title2, "Re: %s", x.title);
			title2[60]=0;
			printf("[<a href='bbspstmail?userid=%s&file=%s&title=%s'>回信</a>]",
				ptr, x.filename, title2);
		}
	}
	fclose(fp);
   	printf("</center>\n"); 
	http_quit();
}

File: kbs_bbs/bbs2www/html/bbsmailcon.php
<?php
	/**
	 * This file displays mail .
	 * $Id$
	 */
	require("funcs.php");
login_init();

	if ($loginok != 1)
		html_nologin();
	else
	{
		if (isset($_GET["num"]))
			$num = $_GET["num"];
		else {
			html_init("gb2312");
			html_error_quit("错误的参数5");
		}

		if (isset($_GET["dir"]))
			$dirname = $_GET["dir"];
		else
			$dirname = ".DIR";

		if (isset($_GET["title"]) )
			$title=$_GET["title"];
		else
			$title="收件箱";
			
		$title_encode = rawurlencode($title);

    	if (strstr($dirname, "..") || strstr($dirname, "/")){
			html_init("gb2312");
			html_error_quit("错误的参数2");
		}
		$dir = "mail/".strtoupper($currentuser["userid"]{0})."/".$currentuser["userid"]."/".$dirname ;

		$total = filesize( $dir ) / 256 ;
		if( $total <= 0 ){
			html_init("gb2312");
			html_error_quit("信笺不存在");
		}

		$articles = array ();
		if( bbs_get_records_from_num($dir, $num, $articles) ) {
			$file = $articles[0]["FILENAME"];
		}else{
			html_init("gb2312");
			html_error_quit("错误的参数4");
		}

		$filename = "mail/".strtoupper($currentuser["userid"]{0})."/".$currentuser["userid"]."/".$file ;
		if(! file_exists($filename)){
			html_init("gb2312");
			html_error_quit("信件不存在...");
		}

			@$attachpos=$_GET["ap"];//pointer to the size after ATTACHMENT PAD
			if ($attachpos!=0) {
				require_once("attachment.php");
				output_attachment($filename, $attachpos);
				exit;
			} else
			{
				html_init("gb2312");
?>
<body>
<center><p><?php echo BBS_FULL_NAME; ?> -- 阅读信件 - <?php echo $title; ?> [使用者: <?php echo $currentuser["userid"]; ?>]</p>
<hr class="default" />
<table width="610" border="0" cellpadding="5" cellspacing="0" class="t1" >
<tr><td class="t5" height="200" valign="top" align="left"><font class="content">
<?php
				bbs_print_article($filename,1,$_SERVER['PHP_SELF'] . '?' . $_SERVER['QUERY_STRING']);
?>
&nbsp;
</font>
</td></tr></table>
<hr>
[<a onclick='return confirm("你真的要删除这封信吗？")' href="/bbsmailact.php?act=del&dir=<?php echo $dirname;?>&file=<?php echo $file;?>&title=<?php echo $title_encode;?>">删除</a>]
[<a href="javascript:history.go(-1)">返回上一页</a>]
<?php
				if($num > 0){
?>
[<a href="/bbsmailcon.php?dir=<?php echo $dirname;?>&num=<?php echo $num-1;?>&title=<?php echo $title_encode;?>">上一篇</a>]
<?php
				}
?>
[<a href="/bbsmailbox.php?path=<?php echo $dirname;?>&title=<?php echo $title_encode;?>&start=<?php echo $num-10;?>">返回信件列表</a>]
<?php
				if($num < $total-1){
?>
[<a href="/bbsmailcon.php?dir=<?php echo $dirname;?>&num=<?php echo $num+1;?>&title=<?php echo $title_encode;?>">下一篇</a>]
<?php
				}
?>
[<a href="/bbspstmail.php?userid=<?php echo $articles[0]["OWNER"]; ?>&file=<?php echo $articles[0]["FILENAME"]; ?>&title=<?php if(strncmp($articles[0]["TITLE"],"Re:",3)) echo "Re: "; ?><?php echo urlencode($articles[0]["TITLE"]); ?>">回信</a>]
</center><br>
<center>
<form action="/bbsmailfwd.php" method="post">
<input type="hidden" name="dir" value="<?php echo $dirname;?>">
<input type="hidden" name="id" value="<?php echo $num;?>">
[把信件转寄给 <input type="text" name="target" size="20" maxlength="69" value="<?php echo $currentuser["userid"];?>"> (输入对方的id或email)
(<input type="checkbox" name="big5" value="1">BIG5)
(<input type="checkbox" name="noansi" value="1">过滤ANSI)
<input type="submit" name="submit" value="转寄">]
</form>
</center>
<?php
			}
		bbs_setmailreaded($dir,$num);
		html_normal_quit();
	}
?>
